### PR TITLE
fix: The DB API Binary function accepts bytes data

### LINKED ISSUE #628: The DB API Binary function should accept bytes.
```
(3.8) jim@ds9:~/p/g/python-bigquery-sqlalchemy$ python
Python 3.8.5 (default, Jan 27 2021, 15:41:15) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import google.cloud.bigquery.dbapi
>>> google.cloud.bigquery.dbapi.Binary(b'x')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jim/p/g/python-bigquery/google/cloud/bigquery/dbapi/types.py", line 42, in Binary
    return string.encode("utf-8")
AttributeError: 'bytes' object has no attribute 'encode'
```
Bytes are the most common way to represent binary data.  Accepting strings, as it does now seems at best to be a convenience and at worst a bug magnet.

In SQLAlchemy, if you defined a model that has a binary attribute, you'd store bytes data in it, but that would break for bigquery, di to this issue.

Sqlite's `Binary` function requires bytes data.

I propose to change the function to accept bytes.  For the sake of backward compatibility, I propose to continue to accept strings.

### PR DESCRIPTION
**Pull request by jimfulton**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #628 🦕

### CODE REVIEW

File: google/cloud/bigquery/dbapi/types.py [hunk 1]
```diff
@@ -30,16 +30,27 @@
...
 
     Returns:
-        bytes: The UTF-8 encoded bytes representing the string.
+        bytes: The binary data as a bytes object.
     """
-    return string.encode("utf-8")
+    if isinstance(data, int):
+        # This is not the conversion we're looking for, because it
+        # will simply create a bytes object of the given size.
+        raise TypeError("cannot convert 'decimal.Decimal' object to binary")
```
[the remark this answers is not in the document]
Comment (jimfulton) [reply google/cloud/bigquery/dbapi/types.py:46]: Oops. Thanks.

[the remark this answers is not in the document]
Comment (jimfulton) [reply google/cloud/bigquery/dbapi/types.py]: Done

File: tests/unit/test_dbapi_types.py [hunk 2]
```diff
@@ -13,6 +13,7 @@
 #  limitations under the License.
 
 import datetime
+import pytest
```
[the remark this answers is not in the document]
Comment (jimfulton) [reply tests/unit/test_dbapi_types.py]: Done

File: tests/unit/test_dbapi_types.py [hunk 3]
```diff
@@ -29,6 +30,19 @@ def test_binary_type(self):
...
+        self.assertEqual(types.Binary(bytearray(b"hello")), b"hello")
+        self.assertEqual(types.Binary(memoryview(b"hello")), b"hello")
+
+        class C:
+            def __bytes__(self):
+                return b"Google"
+
+        self.assertEqual(types.Binary(C()), b"Google")
+
+        for bad in 42, 42.0, None:
```
[the remark this answers is not in the document]
Comment (jimfulton) [reply tests/unit/test_dbapi_types.py]: done

**Review by tswast — APPROVED** [review]

**Pull request merged**